### DISCUSSION TITLE
Libreswan ikev1 quick mode failing.

### DISCUSSION THREAD

**Discussion by gowdamanil** [libreswan/libreswan#1566] [Q&A]
Hello team,

Have started learning about libreswan, I had issues when configuring it. Suggestions for them is greatly welcome.

**### Issue 1:**

**I was trying to setup ipsec between windows (firewall ipsec )  and Linux Libreswan 4.3 (netkey) on 5.10.0-12-amd64.**

I was able to establish main mode but **quick mode is failing.**

For command  **ipsec auto --up myTunnel** 
I am getting output as 

002 "myTunnel" #3: initiating IKEv1 Main Mode connection
102 "myTunnel" #3: sent Main Mode request
002 "myTunnel" #3: WARNING: connection myTunnel PSK length of 13 bytes is too short for HMAC_SHA2_256 PRF in FIPS mode (16 bytes required)
104 "myTunnel" #3: sent Main Mode I2
106 "myTunnel" #3: sent Main Mode I3
002 "myTunnel" #3: Peer ID is ID_IPV4_ADDR: '192.34.x.x'
004 "myTunnel" #3: IKE SA established {auth=PRESHARED_KEY cipher=AES_CBC_128 integ=HMAC_SHA2_256 group=MODP2048}
002 "myTunnel" #4: initiating Quick Mode IKEv1+PSK+ENCRYPT+PFS+UP+IKE_FRAG_ALLOW+ESN_NO {using isakmp#3 msgid:57d1baa8 proposal=AES_CBC_128-HMAC_SHA2_256_128 pfsgroup=MODP2048}
115 "myTunnel" #4: sent Quick Mode request
010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 0.5 seconds for response
010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 1 seconds for response
010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 2 seconds for response
010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 4 seconds for response
010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 8 seconds for response
010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 16 seconds for response
010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 32 seconds for response
031 "myTunnel" #4: STATE_QUICK_I1: 60 second timeout exceeded after 7 retransmits.  No acceptable response to our first Quick Mode message: perhaps peer likes no proposal
000 "myTunnel" #4: starting keying attempt 2 of an unlimited number, but releasing whack

I see there it shows no acceptable response to quick mode message, but on both sides have used the same ciphers.

![ipsecOutput](https://github.com/libreswan/libreswan/assets/153774586/65d74e28-0782-49b1-84fd-fd5a85c79a67)

In the image the algorithms used are: 
 AES_CBC_128 and AES_CBC_256.

The code on libreswan is:

conn myTunnel
        ikev2=never
        type=transport
        authby=secret
        ike=aes128-sha256;modp2048
        esp=aes128-sha256
        left=192.34.x.x
       leftsubnet=192.34.x.x/32
        right=192.34.x.x
       rightsubnet=0.0.0.0/0
        aggressive=no
        salifetime=12h
        ikelifetime=8h
        keyingtries=%forever
        auto=start

 ipsec auto --up mytunnel
000 initiating all conns with alias='mytunnel'
021 no connection named "mytunnel"
root@Phantom-AD8B:~# ipsec auto --up myTunnel
002 "myTunnel" #3: initiating IKEv1 Main Mode connection
102 "myTunnel" #3: sent Main Mode request
104 "myTunnel" #3: sent Main Mode I2
106 "myTunnel" #3: sent Main Mode I3
002 "myTunnel" #3: Peer ID is ID_IPV4_ADDR: '10.x.x.x'
004 "myTunnel" #3: IKE SA established {auth=PRESHARED_KEY cipher=AES_CBC_128 integ=HMAC_SHA2_256 group=MODP2048}
003 "myTunnel" #3: Configured DPD (RFC 3706) support not enabled because remote peer did not advertise DPD support
002 "myTunnel" #4: initiating Quick Mode IKEv1+PSK+ENCRYPT+PFS+UP+IKE_FRAG_ALLOW+ESN_NO {using isakmp#3 msgid:bfe6b66d proposal=3DES_CBC-HMAC_SHA2_256_128-MODP2048, 3DES_CBC-HMAC_SHA1_96-MODP2048, AES_CBC_128-HMAC_SHA2_256_128-MODP2048, AES_CBC_256-HMAC_SHA2_256_128-MODP2048, AES_GCM_16-NONE-MODP2048 pfsgroup=MODP2048}
115 "myTunnel" #4: sent Quick Mode request
010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 0.5 seconds for response
010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 1 seconds for response
010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 2 seconds for response
010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 4 seconds for response
010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 8 seconds for response
010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 16 seconds for response
010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 32 seconds for response
031 "myTunnel" #4: STATE_QUICK_I1: 60 second timeout exceeded after 7 retransmits.  No acceptable response to our first Quick Mode message: perhaps peer likes no proposal
000 "myTunnel" #4: starting keying attempt 2 of an unlimited number, but releasing whack

Please see the ciphers configured in above image for server and in libreswan code, they are same even then i see failing of quick mode. 

Thank you 
I know modp2048 is old but needed for compatibility.

Thanks

**Comment by cagney** [comment]
```
when i used esp=aes128-sha256, aes256-sha256

036 "myTunnel": failed to add connection: ESP encryption algorithm '3des+aes128' is not recognized
```
this suggests that `esp=3des+aes128` was tried, which is IKEv2 syntax.

**Reply by gowdamanil** [reply]
Hello cagney,

I was able to get main mode request and was succesfully established, but quick mode is failing as shown in the above question.

 ipsec auto --up mytunnel
000 initiating all conns with alias='mytunnel'
021 no connection named "mytunnel"
root@Phantom-AD8B:~# ipsec auto --up myTunnel
002 "myTunnel" #3: initiating IKEv1 Main Mode connection
102 "myTunnel" #3: sent Main Mode request
104 "myTunnel" #3: sent Main Mode I2
106 "myTunnel" #3: sent Main Mode I3
002 "myTunnel" #3: Peer ID is ID_IPV4_ADDR: '10.x.x.x'
004 "myTunnel" #3: IKE SA established {auth=PRESHARED_KEY cipher=AES_CBC_128 integ=HMAC_SHA2_256 group=MODP2048}
003 "myTunnel" #3: Configured DPD (RFC 3706) support not enabled because remote peer did not advertise DPD support
002 "myTunnel" #4: initiating Quick Mode IKEv1+PSK+ENCRYPT+PFS+UP+IKE_FRAG_ALLOW+ESN_NO {using isakmp#3 msgid:bfe6b66d proposal=3DES_CBC-HMAC_SHA2_256_128-MODP2048, 3DES_CBC-HMAC_SHA1_96-MODP2048, AES_CBC_128-HMAC_SHA2_256_128-MODP2048, AES_CBC_256-HMAC_SHA2_256_128-MODP2048, AES_GCM_16-NONE-MODP2048 pfsgroup=MODP2048}
115 "myTunnel" #4: sent Quick Mode request
010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 0.5 seconds for response
010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 1 seconds for response
010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 2 seconds for response
010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 4 seconds for response
010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 8 seconds for response
010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 16 seconds for response
010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 32 seconds for response
031 "myTunnel" #4: STATE_QUICK_I1: 60 second timeout exceeded after 7 retransmits.  No acceptable response to our first Quick Mode message: perhaps peer likes no proposal
000 "myTunnel" #4: starting keying attempt 2 of an unlimited number, but releasing whack

Please refer the image in my question of windows configuration for ipsec vs libreswan, both are same even then i am failing to establish connection.

Regards,
Anil Gowda M

**Comment by letoams** [comment]
On Tue, 27 Feb 2024, gowdamanil wrote:

> Thank you but can you guys give me syntax for ikev1

 	esp=aes128-sha2,aes256-sha2,3des-sha1

(ofcourse best to not use 3des or sha1 at all)

> On Mon, 26 Feb, 2024, 8:03 pm cagney, ***@***.***> wrote:
> 
> > when i used esp=aes128-sha256, aes256-sha256
> >
> > 036 "myTunnel": failed to add connection: ESP encryption algorithm '3des+aes128' is not recognized

That esp= line cannot throw that ESP error ling.

> > this suggests that esp=3des+aes128 was tried, which is IKEv2 syntax.

This might be coming in via crypto policies? But that should only carry
onwards in esp= was not set at all for the connection.

**Reply by gowdamanil** [reply]
Hello @letoams,

Thank you for reply, I have checked my both systems, I was able to bring up phase1.Also libreswan is sending quick mode request, but is not receiving any responses from windows ipsec(not even no matching ciphers). That's my main issue.

Here is the output of ipsec command:   **ipsec auto --up mytunnel** 


 002 "myTunnel" https://github.com/libreswan/libreswan/issues/4: initiating Quick Mode IKEv1+PSK+ENCRYPT+PFS+UP+IKE_FRAG_ALLOW+ESN_NO {using isakmp#3 msgid:bfe6b66d proposal=3DES_CBC-HMAC_SHA2_256_128-MODP2048, 3DES_CBC-HMAC_SHA1_96-MODP2048, AES_CBC_128-HMAC_SHA2_256_128-MODP2048, AES_CBC_256-HMAC_SHA2_256_128-MODP2048, AES_GCM_16-NONE-MODP2048 pfsgroup=MODP2048}
115 "myTunnel" https://github.com/libreswan/libreswan/issues/4: sent Quick Mode request
010 "myTunnel" https://github.com/libreswan/libreswan/issues/4: STATE_QUICK_I1: retransmission; will wait 0.5 seconds for response
010 "myTunnel" https://github.com/libreswan/libreswan/issues/4: STATE_QUICK_I1: retransmission; will wait 1 seconds for response
010 "myTunnel" https://github.com/libreswan/libreswan/issues/4: STATE_QUICK_I1: retransmission; will wait 2 seconds for response
010 "myTunnel" https://github.com/libreswan/libreswan/issues/4: STATE_QUICK_I1: retransmission; will wait 4 seconds for response
010 "myTunnel" https://github.com/libreswan/libreswan/issues/4: STATE_QUICK_I1: retransmission; will wait 8 seconds for response
010 "myTunnel" https://github.com/libreswan/libreswan/issues/4: STATE_QUICK_I1: retransmission; will wait 16 seconds for response
010 "myTunnel" https://github.com/libreswan/libreswan/issues/4: STATE_QUICK_I1: retransmission; will wait 32 seconds for response
031 "myTunnel" https://github.com/libreswan/libreswan/issues/4: STATE_QUICK_I1: 60 second timeout exceeded after 7 retransmits. No acceptable response to our first Quick Mode message: perhaps peer likes no proposal
000 "myTunnel" https://github.com/libreswan/libreswan/issues/4: starting keying attempt 2 of an unlimited number, but releasing whack

**I need to succesfully establish phase2.**

**If you need steps I followed for setting up in windows. Have attached here:-**
	
1. Firewall should be enabled.

Control Panel > System and Security > Windows Defender Firewall > Turn windows defender firewall on or off > Firewall should be enabled for private and public network settings
	
2. Firewall state should be ON for the profile with which the client communicates with libreswan(domain/public/private profiles).

Open Windows 10 run box (WIN R) >  gpedit.msc  > computer configuration > windows settings > security settings > windows defender firewall with advanced security > right click > properties > enable the firewall for required profile



**Steps to configure IPSEC using Pre-shared key-based authentication:**


1. Open Windows 10 run box (WIN R) >  gpedit.msc > Computer Configuration > Windows Settings > Security Settings > Windows Defender Firewall with Advanced Security > Windows Defender Firewall with Advanced Security- Local Group Policy Object > right-click on connection security rules
	
2. Click on new rule > Choose the rule type which is appropriate to the environment (we will add custom rule here.) > click Next

3.  Select any IP address in Endpoint 1(My windows machine itself where i am configuring ipsec inbuilt) and in Endpoint 2 (libreswan configured) and  click OK and click Next	
	
4. In Requirements, choose the option which is appropriate for your environment and click Next

5. Authentication method > Advanced > click on customize > add preshared key > click OK  >click Next 	

6. Protocol and ports > click Next	

7. Profiles > click Next	

8. Provide the name > click on FInish	
	
Changing the ciphers(Optional)

9. Open Windows 10 run box (WIN R) >  gpedit.msc > Computer Configuration > Windows Settings > Security Settings > Windows Defender Firewall with Advanced Security > Windows Defender Firewall with Advanced Security- Local Group Policy Object > right-click > Properties > IPSec settings > IPSec defaults > customize > Key exchange (main mode) > Advanced > Customize > Click on Add > Select Integrity, Encryption and Key exchange algorithms from drop down > Click OK 

10. Data protection (quick mode) > Advanced > Click on Add > Select Integrity, Encryption  algorithms from drop down > Click OK 

(Configure same ciphers on bothsides).

Note: With the above configuration, the default ciphers used are SHA 1 for Integrity and AES-CBC 128 for Encryption. In case, the user wants to modify the ciphers, user has to refer step 9


  

I tried same with strongswan and both phases were correctly accepted and established, but in libreswan failing.

This is an feature that i am researching about since 2 months and I will gain in depth knowledge as i proceed further so please notify about my mistakes and steps that i should take for it to rectify them.

Please ask if any more info is needed to debug this.

Thanks,
Anil GM

**Comment by letoams** [comment]
On Thu, 29 Feb 2024, gowdamanil wrote:

> I have checked my both systems, I was able to bring up phase1 but libreswan is sending quick mode
> request, but is not receiving any responses from windows ipsec(not even no matching ciphers).
> That's my main issue.
> 
> 002 "myTunnel" #4: initiating Quick Mode IKEv1+PSK+ENCRYPT+PFS+UP+IKE_FRAG_ALLOW+ESN_NO {using
> isakmp#3 msgid:bfe6b66d proposal=3DES_CBC-HMAC_SHA2_256_128-MODP2048,
> 3DES_CBC-HMAC_SHA1_96-MODP2048, AES_CBC_128-HMAC_SHA2_256_128-MODP2048,
> AES_CBC_256-HMAC_SHA2_256_128-MODP2048, AES_GCM_16-NONE-MODP2048 pfsgroup=MODP2048}
> 115 "myTunnel" #4: sent Quick Mode request
> 010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 0.5 seconds for response
> 010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 1 seconds for response
> 010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 2 seconds for response
> 010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 4 seconds for response
> 010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 8 seconds for response
> 010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 16 seconds for response
> 010 "myTunnel" #4: STATE_QUICK_I1: retransmission; will wait 32 seconds for response
> 031 "myTunnel" #4: STATE_QUICK_I1: 60 second timeout exceeded after 7 retransmits. No acceptable
> response to our first Quick Mode message: perhaps peer likes no proposal
> 000 "myTunnel" #4: starting keying attempt 2 of an unlimited number, but releasing whack

What did the other end say? Did it receive a STATE_QUICK_I1 packet and
log an error? If not, it seems to be a network issue. What size are
those STATE_QUICK_I1 packets? I think the log should tell you the packet
size. Perhaps it is too close to your MTU size and you have some bad
firewall / router on your path?

> I tried same with strongswan and both phases were correctly accpeted and established, but in
> libreswan failing.

that would seem to suggest no network problems, whih means the end
receiving the STATE_QUICK_I1 packets should log why it is not
responding.

**Reply by gowdamanil** [reply]
> that would seem to suggest no network problems, whih means the end
> receiving the STATE_QUICK_I1 packets should log why it is not
> responding.

I saw STATE_QUICK_l1 packet being received on windows side, but no response is being given from windows.
On windows side, I have not seen events on that side(Windows event viewer). Maybe logs are stored elsewhere.
I have checked them extensively and was unable to find logs for that event.On libreswan i see no messages but just retransmition.

**Comment by letoams** [comment]
On Mon, 4 Mar 2024, gowdamanil wrote:

> I saw STATE_QUICK_l1 packet being received on windows side, but no response is being given from windows.

That suggests some of the phase 2 parameters is not acceptable to it.

- check left/right and leftsubnet/rightsubnet
- try changing pfs=  (yes to no or no to yes)
- check esp= line and try other/more algorithms to see if any work?

> On windows side, I have not seen events on that side(Windows event viewer). Maybe logs are stored elsewhere.
> I have checked them extensively and was unable to find logs for that event.On libreswan i see no messages but just
> retransmition.

There used to be some regedit value to enable ipsec debugging.

Paul

**Reply by gowdamanil** [reply]
Hello @letoams 

> That suggests some of the phase 2 parameters is not acceptable to it. - check left/right and leftsubnet/rightsubnet - try changing pfs= (yes to no or no to yes) - check esp= line and try other/more algorithms to see if any work?
> On windows side, I have not seen events on that side(Windows event viewer). Maybe logs are stored elsewhere. I have checked them extensively and was unable to find logs for that event.On libreswan i see no messages but just retransmition.


I tried multiple algorithms available for esp, which were (AES-GMAC 128 , 192, 256. Also AES-GCM 128, 192, 256. )None of them was able to bring up 2nd phase.pfs also was of no use. left, rightsubnet was also used correctly.

> There used to be some regedit value to enable ipsec debugging. 


HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\PolicyAgent\Logging\Oakley
For EnableLogging (DWORD), set it to 1 to enable logging.
For LogFileMaxSize, enter the maximum log file size in bytes (for example, 50000).
For LogLevel, choose a value from 0 (least detailed) to 4 (most detailed).
After that restarted policy agent and stopped ipsec and again restarted it.Even then I was unable to find log file in %SystemRoot%\System32\LogFiles\IPsec\

**Comment by letoams** [comment]
> On Mar 18, 2024, at 22:14, gowdamanil ***@***.***> wrote:
> 
> I tried multiple algorithms available for esp, which were (AES-GMAC 128 , 192, 256. Also AES-GCM 128, 192, 256. )None of them was able to bring up 2nd phase.
> 

Those are not common for ikev1. Try aes256-sha1, aes128-sha1,aes256-sha2  which will use AES-CBC.


> There used to be some regedit value to enable ipsec debugging.
> 
> HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\PolicyAgent\Logging\Oakley
> For EnableLogging (DWORD), set it to 1 to enable logging.
> For LogFileMaxSize, enter the maximum log file size in bytes (for example, 50000).
> For LogLevel, choose a value from 0 (least detailed) to 4 (most detailed).
> After that restarted policy agent and stopped ipsec and again restarted it.Even then I was unable to find log file in %SystemRoot%\System32\LogFiles\IPsec\
> 

I think it requires a reboot
